src/main/java/com/redhat/devtools/lsp4ij/features/formatting/LSPClientSideOnTypeFormattingTypedHandler.java

```diff
@@ -0,0 +1,277 @@
+        for (LanguageServerWrapper startedLanguageServer : startedLanguageServers) {
+            // TODO: Is there a better way to ask if this language server supports the file?
+            if (startedLanguageServer.isConnectedTo(LSPIJUtils.toUri(file))) {
+                LSPFormattingFeature formattingFeature = startedLanguageServer.getClientFeatures().getFormattingFeature();
+                if (formattingFeature.isEnabled(file) && formattingFeature.isSupported(file)) {
+                    // TODO: This returns the first match. Is that okay?
+                    return formattingFeature;
+                }
+            }
+        }
+        return null;
+    }
+
+    @NotNull
+    private static Result handleCloseBraceTyped(@NotNull Project project,
+                                                @NotNull Editor editor,
+                                                @NotNull PsiFile file,
+                                                @NotNull LSPFormattingFeature formattingFeature,
+                                                char openBraceChar,
+                                                char closeBraceChar) {
+        TextRange formatTextRange = null;
+
+        // Statement-level scope is not supported for code blocks
+        FormattingScope formattingScope = formattingFeature.getFormatOnCloseBraceScope(file);
+        if (formattingScope == FormattingScope.STATEMENT) {
+            return Result.CONTINUE;
+        }
+
+        // If appropriate, find the code block that was closed by the brace
+        if (formattingScope == FormattingScope.CODE_BLOCK) {
+            int offset = editor.getCaretModel().getOffset();
+            int beforeOffset = offset - 1;
+            TextRange codeBlockRange = LSPCodeBlockProvider.getCodeBlockRange(editor, file, beforeOffset);
+            if (codeBlockRange != null) {
+                int startOffset = codeBlockRange.getStartOffset();
+                int endOffset = codeBlockRange.getEndOffset();
+
+                // Make sure the range includes the brace pair
+                Document document = editor.getDocument();
+                CharSequence documentChars = document.getCharsSequence();
+                if ((startOffset > 0) && (documentChars.charAt(startOffset) != openBraceChar)) {
+                    startOffset--;
+                }
+                if ((endOffset < (documentChars.length() - 1)) && (documentChars.charAt(endOffset) != closeBraceChar)) {
+                    endOffset++;
+                }
+
+                // If the range is now the braced block, format it
+                if ((documentChars.charAt(startOffset) == openBraceChar) && (documentChars.charAt(endOffset) == closeBraceChar)) {
+                    formatTextRange = TextRange.create(startOffset, endOffset);
+                }
+            }
+        }
+
+        // If appropriate, use the file text range
+        else if (formattingScope == FormattingScope.FILE) {
```

Comment:
This can be configured for file scope if desired, but if the language server provides enough information for reliable code blocks, it should be configured for code block scope (which is the default, of course).